zerver/actions/message_flags.py

```diff
@@ -262,6 +263,28 @@ def do_update_message_flags(
     flagattr = getattr(UserMessage.flags, flag)
 
     with transaction.atomic(savepoint=False):
+        if flag == "read" and operation == "remove":
+            # We have an invariant that all stream messages marked as
+            # unread must be in streams the user is subscribed to.
+            #
+            # When marking as unread, we enforce this invariant by
+            # ignoring any messages in streams the user is not
+            # currently subscribed to.
+            subscribed_recipient_ids = get_subscribed_stream_recipient_ids_for_user(user_profile)
+
+            message_ids_in_unsubscribed_streams = (
+                Message.objects.select_related("recipient")
+                .filter(id__in=messages, recipient__type=Recipient.STREAM)
+                .exclude(recipient_id__in=subscribed_recipient_ids)
+                .values_list("id", flat=True)
+            )
+
+            messages = [
+                message_id
+                for message_id in messages
+                if message_id not in set(message_ids_in_unsubscribed_streams)
+            ]
```

Comment:
I believe this is quadratic: the `if`-condition will get evaluated once for each iteration of the `for`. So we'll construct a new `set` of all the excluded message IDs once for each element of `messages`.

I think a sufficient fix is to say
```py
  message_ids_in_unsubscribed_streams = set(
    # … same query …
  )
```
above, instead of calling `set` here.

Reply:
Good catch; fixed in 873ab01c7a534dd7169d07112b320d51c5592da7.